File: docs/JAMIA/comparisons/RPython/CDA/38516481_replace_vs_str_replace.py
"""
RPython SO#38516481 — Series.replace() silently does nothing on substrings
Effect: IF (silent incorrect functionality)

pandas Series.replace(',', '') does VALUE-level replacement (looking for cells
that are exactly ','). It does NOT do substring replacement within cells.
The column is unchanged — no error, no warning. You need .str.replace() instead.
"""
import pandas as pd

df = pd.DataFrame({
    'Player': ['Mike Trout', 'Clayton Kershaw', 'Bryce Harper'],
    'Avg_Annual': ['$34,083,333', '$31,000,000', '$25,538,462'],
})

# BUG: .replace() silently does nothing — looks for exact cell match
df['Avg_Annual'] = df['Avg_Annual'].replace(',', '')
df['Avg_Annual'] = df['Avg_Annual'].replace('$', '')

print("After .replace() — UNCHANGED (silent failure):")
print(df['Avg_Annual'].tolist())
# ['$34,083,333', '$31,000,000', '$25,538,462']
# No error. User thinks they cleaned the data. They didn't.
print()

# Correct: .str.replace() for substring operations
df['Avg_Annual'] = df['Avg_Annual'].str.replace(',', '', regex=False)
df['Avg_Annual'] = df['Avg_Annual'].str.replace('$', '', regex=False)
print("After .str.replace() — correct:")
print(df['Avg_Annual'].tolist())

File: docs/JAMIA/comparisons/RPython/CDA/38516481_replace_vs_str_replace.ts
/**
 * RPython SO#38516481 — Series.replace() silently does nothing on substrings
 * Effect: IF (silent incorrect functionality)
 * Bug class: API ambiguity
 *
 * In pandas, Series.replace(',', '') does VALUE-level replacement (exact cell match).
 * It silently does nothing when the user wants substring replacement.
 * The user proceeds thinking the data was cleaned.
 *
 * In tidy-ts, string manipulation in mutate() uses JS string methods.
 * .replaceAll() operates on substrings. The column is typed as string,
 * so the user must explicitly convert to number before numeric operations.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const df = createDataFrame([
  { player: "Mike Trout", avg_annual: "$34,083,333" },
  { player: "Clayton Kershaw", avg_annual: "$31,000,000" },
  { player: "Bryce Harper", avg_annual: "$25,538,462" },
]);

// avg_annual is string — attempting numeric operations without cleaning fails
// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.mean(df.extract("avg_annual"));

// Clean and convert
const cleaned = df.mutate({
  salary: (r) => parseFloat(r.avg_annual.replaceAll(",", "").replaceAll("$", "")),
});

console.log(s.mean(cleaned.extract("salary")));
